ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,341 @@
+    this.aliveHosts = new ConcurrentHashMap<>();
+    this.remoteHostsURI = new ConcurrentHashMap<>();
+    this.remoteHostsInfo = new ConcurrentHashMap<>();
+  }
+
+  void scheduleHeartbeatSender() {
+    SendHeartbeat sendTask = new SendHeartbeat();
+    scheduledExecutorService.scheduleAtFixedRate(sendTask, 0, heartbeatSendInterval,
+                                                 TimeUnit.MILLISECONDS);
+  }
+
+  void scheduleHeartbeatProcessor() {
+    ProcessHeartbeat processTask = new ProcessHeartbeat();
+    scheduledExecutorService.scheduleAtFixedRate(processTask, 0, heartbeatCheckInterval,
+                                                 TimeUnit.MILLISECONDS);
+  }
+
+  void scheduleClusterDiscoverer() {
+    DiscoverCluster discoverTask = new DiscoverCluster();
+    scheduledExecutorService.scheduleAtFixedRate(discoverTask, 0, discoverClusterInterval,
+                                                 TimeUnit.MILLISECONDS);
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  @VisibleForTesting
+  void initialize() {
+    aliveHosts.clear();
+    remoteHostsInfo.clear();
+    remoteHostsURI.clear();
+    receivedHeartbeats.clear();
+  }
+
+  void buildLocalHostInfo(final URI localServerURL) {
+
+    try {
+      localHostInfo = new HostInfo(localServerURL.getHost(), localServerURL.getPort());
+      localURL = localServerURL.toURL();
+      aliveHosts.putIfAbsent(localHostInfo.toString(), true);
+    } catch (final Exception e) {
+      throw new IllegalStateException("Failed to convert local host URI to URL."
+                                          + " local host URL: " + localServerURL);
+    }
+  }
+
+  class ProcessHeartbeat implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - heartbeatWindow;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        //Question: Until this thread gets the lock, "now" will have changed. But if I compute "now"
```

Comment:
discussed.. may be better to leave a GH comment for this in the future?  so you dont have have change code to remove the question? :)

Reply:
You mean when I do the actual PR, to put this question as a comment ?